Alex and Happy Part 1/article.md

```diff
@@ -0,0 +1,768 @@
+For example, here's how we can match an identifier:
+
+```alex
+$digit = [0-9]
+$alpha = [a-zA-Z]
+
+@id = ($alpha | \_) ($alpha | $digit | \_ | \' | \?)*
+
+tokens :-
+```
+
+We use `@NAME = REGEX` to define a regular expression macro. Note that macros may not be recursive.
+
+#### Lexing identifiers
+
+Let's define our first lexeme based on our identifier:
+
+```alex
+tokens :-
+
+<0> $white+ ;
+
+<0> @id     { tokId }
+```
+The syntax `<START_CODE> REGEX { CODE }` means that if Alex has managed to match a token `REGEX`, and it's currently in the `START_CODE` state, it should execute the code `CODE`.
+
+Similarly, for white space, we can use `;` instead of giving it an explicit action, which simply means that Alex should do nothing interesting with it. This is the same as writing `{ skip }`.
+
+The idea of start codes is a bit more sophisticated yet valuable.
+Alex works as a state machine, and `0` indicates the initial code in which the machine starts.
+In this case, we can match white space and identifiers when we start in state `0`.
+We will later create other start codes.
+
+`CODE` may contain any Haskell expression, as it will be included verbatim in the generated code.
+Alex expects that this expression will have type `AlexAction RangedToken`. Here's the automatically-genereated definition for `AlexAction` when using `monadUserState-bytestring`:
+
+```haskell
+type AlexAction result = AlexInput -> Int64 -> Alex result
+```
+
+Where `result` will be `RangedToken` in our case.
+
+Let's create our `tokId` function. We do so by inserting this definition at the bottom of the file:
+
+```haskell
+mkRange :: AlexInput -> Int64 -> Range
+mkRange (start, _, str, _) len = Range{start = start, stop = stop}
+  where
+    stop = BS.foldl' alexMove start $ BS.take len str
+
+tokId :: AlexAction RangedToken
+tokId inp@(_, _, str, _) len =
+  pure RangedToken
+    { rtToken = Identifier $ BS.take len str
+    , rtRange = mkRange inp len
+    }
+```
+
+`tokId` will extract the lexeme from the input string, which will be the first `len` characters of `str`.
+The range will then be handled by `mkRange`, a function that advances the position accordingly to each seen character in the input string using `alexMove`. Alex automatically generates this function, and you don't need to include it with your code.
+
+We should check whether our code works. I will define a small function that we can use for testing. Don't forget to export it!
+
+```haskell
+scanMany :: ByteString -> Either String [RangedToken]
+scanMany input = runAlex input go
+  where
+    go = do
+      output <- alexMonadScan
+      if rtToken output == EOF
+        then pure [output]
+        else (output :) <$> go
+```
+
+Startup GHCi and let's check whether this works. It should result in something like this (slightly prettified for ease of visualization):
+
+```haskell
+>>> runAlex "my_identifier" alexMonadScan
```

Comment:
I put a few more words about Stack. Please let me know if after updating Stack and using `stack ghci` you still have difficulties running these examples.

Just so that we don't forget, during the call, you suggested placing more anchors in the text, which I've done as well.

Reply:
Gist checkpoints could be helpful, perhaps?

Reply:
I will add a few more checkpoints later.